x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/oss/IndexUsageTransportAction.java

```diff
@@ -0,0 +1,122 @@
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.metadata.MappingMetaData;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.protocol.xpack.XPackUsageRequest;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.action.XPackUsageFeatureAction;
+import org.elasticsearch.xpack.core.action.XPackUsageFeatureResponse;
+import org.elasticsearch.xpack.core.action.XPackUsageFeatureTransportAction;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class IndexUsageTransportAction extends XPackUsageFeatureTransportAction {
+
+    @Inject
+    public IndexUsageTransportAction(TransportService transportService, ClusterService clusterService, ThreadPool threadPool,
+                                     ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(XPackUsageFeatureAction.INDEX.name(), transportService, clusterService, threadPool, actionFilters,
+            indexNameExpressionResolver);
+    }
+
+    @Override
+    protected void masterOperation(Task task, XPackUsageRequest request, ClusterState state,
+                                   ActionListener<XPackUsageFeatureResponse> listener) {
+
+        final Set<String> usedFieldTypes = new HashSet<>();
+        final Set<String> usedCharFilters = new HashSet<>();
+        final Set<String> usedTokenizers = new HashSet<>();
+        final Set<String> usedTokenFilters = new HashSet<>();
+        final Set<String> usedAnalyzers = new HashSet<>();
+
+        for (IndexMetaData indexMetaData : state.metaData()) {
+            MappingMetaData mappingMetaData = indexMetaData.mapping();
+            if (mappingMetaData != null) {
+                populateFieldTypesFromObject(mappingMetaData.sourceAsMap(), usedFieldTypes);
+            }
+
+            Settings indexSettings = indexMetaData.getSettings();
+
+            Map<String, Settings> charFilterSettings = indexSettings.getGroups("index.analysis.char_filter");
+            aggregateAnalysisTypes(charFilterSettings.values(), usedCharFilters);
+
+            Map<String, Settings> tokenizerSettings = indexSettings.getGroups("index.analysis.tokenizer");
+            aggregateAnalysisTypes(tokenizerSettings.values(), usedTokenizers);
+
+            Map<String, Settings> tokenFilterSettings = indexSettings.getGroups("index.analysis.filter");
+            aggregateAnalysisTypes(tokenFilterSettings.values(), usedTokenFilters);
+
+            Map<String, Settings> analyzerSettings = indexSettings.getGroups("index.analysis.analyzer");
+            aggregateAnalysisTypes(analyzerSettings.values(), usedAnalyzers);
```

Comment:
Should we also extract the pre-built tokenizer, filters and char_filters from custom analyzers ?